### PR TITLE
FALCON-1853 : Add spark process workflow builder

### PR DESCRIPTION
**Pull request by peeyushb**
Following is the pull request that will perform the integration to execute the Spark application through Falcon. Please review. 

XSD changes for integration are already committed through pull request #89 .

### CODE REVIEW

**Comment by peeyushb** [comment]
I have verified the Falcon build using this pull request on secure cluster and it works successfully. Spark process entity runs with no issues.

File: oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+            Property sparkProp = new Property();
+            sparkProp.setName(prop.getName());
+            sparkProp.setValue(prop.getValue());
+            sparkConf.add(sparkProp);
+        }
+    }
+
+    private void checkDataFeedsAsArgument(List<String> argList, Cluster cluster) throws FalconException {
+        if (entity.getInputs() == null) {
+            LOG.error("Input Feed Entity is null for process : {}" +entity.getName());
```
Comment (vrangan) [inline oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java]: Why do we mandate that input/output feeds must be present

File: oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            return;
+        }
+
+        if (entity.getOutputs() == null) {
+            LOG.error("Output Feed Entity is null for process : {}" +entity.getName());
```
Comment (vrangan) [inline oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java]: Why do we mandate that input/output feeds must be present?

File: oozie/src/test/java/org/apache/falcon/oozie/process/OozieProcessWorkflowBuilderTest.java [hunk 3]
```diff
@@ -88,6 +88,7 @@
     private static final String FEED_XML = "/config/feed/feed-0.1.xml";
     private static final String CLUSTER_XML = "/config/cluster/cluster-0.1.xml";
     private static final String PIG_PROCESS_XML = "/config/process/pig-process-0.1.xml";
+    private static final String SPARK_PROCESS_XML = "/config/process/spark-process.xml";
```
Comment (vrangan) [inline oozie/src/test/java/org/apache/falcon/oozie/process/OozieProcessWorkflowBuilderTest.java]: Can we add the 0.1 versioning to spark-process xml also

File: oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java (hunk 1 above)
Comment (peeyushb) [inline oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java]: The purpose of mandating the input/output feeds(which has to be passed as argument to Spark application in process entity) to ensure that spark application must read data and write data from/to Falcon specified Feed entities. Moreover I have checked the pig/hive process entity examples and found that input/output to these examples are based on already defined feed entity only.

Comment (peeyushb) [inline oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java]: I am of the opinion that let the user specify about the name of the input/output feed entity as argument to spark application and I will remove the input/output checks. 

We will document with examples that parameter from input/output feed entity must be passed as arguments to the Spark application.

File: oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java (hunk 2 above)
Comment (peeyushb) [inline oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java]: removed

**Comment by peeyushb** [comment]
Thanks @vrangan for reviewing the pull request. Pull request has been updated after incorporating the review comments.

**Comment by vrangan** [comment]
[INFO] Method org.apache.falcon.oozie.process.SparkProcessWorkflowBuilder.checkDataFeedsAsArgument(Process) seems to be useless [org.apache.falcon.oozie.process.SparkProcessWorkflowBuilder] At SparkProcessWorkflowBuilder.java:[line 144] UC_USELESS_VOID_METHOD

**Comment by vrangan** [comment]
Can you fix this and make sure you run complete fix and test with the latest master and use
-Ptest-patch,hivedr,hdfs-snapshot-mirroring,adf to make sure we have all components are covered

**Comment by peeyushb** [comment]
@vrangan I have fixed the issue and now build is successful. Also I have tested the patch it is working.

**Comment by peeyushb** [comment]
I have checked the jenkins build failure and they are not related to this patch.

**Comment by vrangan** [comment]
@peeyushb   - The patch no longer applies.  Can you please rebase and reupload

**Comment by vrangan** [comment]
It looks like multicommit pull requests don't apply if we get a patch - works ok if we do the final diff

File: oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        String sparkJobName = entity.getSparkAttributes().getName();
+        String sparkOpts = entity.getSparkAttributes().getSparkOpts();
+        String sparkClassName = entity.getSparkAttributes().getClazz();
+
+        String clusterEntitySparkMasterURL = getClusterEntitySparkMaster(cluster);
+
+        //Overriding cluster spark master url if defined in process entity
+        sparkMasterURL = (sparkMasterURL == null) ? clusterEntitySparkMasterURL : sparkMasterURL;
+        if (StringUtils.isBlank(sparkMasterURL)) {
+            LOG.error("Spark Master URL can'be empty");
```
Comment (pavankumar526) [inline oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java]: Shouldn't we throw an exception here ?

File: oozie/src/main/resources/action/process/spark-action.xml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+                <value>${queueName}</value>
+            </property>
+            <property>
+                <name>oozie.launcher.mapred.job.priority</name>
+                <value>${jobPriority}</value>
+            </property>
+        </configuration>
+        <master>#MASTER_URL#</master>
+        <name>#NAME#</name>
+        <jar>#JAR_FILE_PATH</jar>
```
Comment (pavankumar526) [inline oozie/src/main/resources/action/process/spark-action.xml]: # missing

File: common/pom.xml [hunk 6]
```diff
@@ -207,6 +207,31 @@
...
+                    <groupId>com.fasterxml.jackson.module</groupId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.eclipse.jetty.orbit</groupId>
+                    <artifactId>javax.servlet</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+
+        <dependency>
```
Comment (pavankumar526) [inline common/pom.xml]: Just want to know why is this required ?

Comment (peeyushb) [inline common/pom.xml]: When I added the spark-core artifact in common module, one of the test case failed with "NoClassDefFoundError org/jets3t/service/ServiceException", so I added this artifact to avoid failure.

File: oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java (hunk 4 above)
Comment (peeyushb) [inline oozie/src/main/java/org/apache/falcon/oozie/process/SparkProcessWorkflowBuilder.java]: Fixed.

File: common/pom.xml (hunk 6 above)
Comment (peeyushb) [inline common/pom.xml]: When I added the spark-core artifact in common module, one of the test case failed with "NoClassDefFoundError org/jets3t/service/ServiceException", so I added this artifact to avoid failure.

**Comment by peeyushb** [comment]
Thanks @pavankumar526  for reviewing the pull request. I have incorporated the review comments.

**Comment by vrangan** [comment]
+1

Thanks @peeyushb  for diligently working on this and responding to comments.

Can you check the one failure in jenkins runs - It does not look related - just review and comment

**Comment by peeyushb** [comment]
Thanks @vrangan for reviewing the patch in pull request. 

I have done a check for failure in Jenkins and it does not look related.

File: common/pom.xml (hunk 6 above)
Comment (pavankumar526) [inline common/pom.xml]: Can we scope this dependency to tests ?

Comment (peeyushb) [inline common/pom.xml]: Done

**Comment by pavankumar526** [comment]
+1 patch looks good. @peeyushb It will be good to have one integration test using Falcon Unit, so that it will cover end to end. You can do it as part of other jira.

**Comment by pavankumar526** [comment]
Will merge this patch in some time, if there are no further comments
